### DISCUSSION TITLE
Best way to switch to maintenance mode

### DISCUSSION THREAD

**Discussion by visakadivines** [vercel/next.js#12850] [Help] [answered]
I’ve created a maintenance page that I’d like all routes to redirect to for a specified day/time every week. The way I see it working right now is to have a separate now.json, or modify the existing one, where I redirect all routes to the maintenance page and set up a deploy hook to redeploy the website. Just wondering if there are any best practices when it comes to making updates to the now.json file and redeploying? Are others taking a similar approach or is there a better way to go about this?

**Comment by leerob** [comment]
You could add a check in `_app.js` for maintenance mode and then conditionally change which page you show. 

https://nextjs.org/docs/advanced-features/custom-app

**Reply by runoncedev** [reply]
What if my backend is down? I can't even build my Next.js app because every request made from `getStaticProps` will fail.

I'm trying to use a redirect but is not working properly, even in the most basic form:

```
module.exports = {
  async redirects() {
    return [
      {
        source: "/",
        destination: "/maintenance",
        permanent: false,
      },
    ];
  },
};
````

This will only redirect on the first page load for some reason.

For example:

 * The maintenance page includes a link to the homepage (because it's part of the main layout)
 * The user clicks the link to the homepage
 * It gets redirected to `/`
 *  The 404 page is rendered.

I was expcting that the user would be redirected back to the maintenance page after trying to navigate to the homepage.

**Reply by hoangvu12** [reply]
> What if my backend is down? I can't even build my Next.js app because every request made from `getStaticProps` will fail.
> 
> I'm trying to use a redirect but is not working properly, even in the most basic form:
> 
> ```
> module.exports = {
>   async redirects() {
>     return [
>       {
>         source: "/",
>         destination: "/maintenance",
>         permanent: false,
>       },
>     ];
>   },
> };
> ```
> 
> This will only redirect on the first page load for some reason.
> 
> For example:
> 
> * The maintenance page includes a link to the homepage (because it's part of the main layout)
> * The user clicks the link to the homepage
> * It gets redirected to `/`
> * The 404 page is rendered.
> 
> I was expcting that the user would be redirected back to the maintenance page after trying to navigate to the homepage.

What I did is create another branch, then put a index.html file that show the maintance alert to the user. Then publish deployed branch whenever I want (I'm using Netlify so I can publish old deploy).

**Reply by runoncedev** [reply]
I did something similar, using a separated branch but instead of an index.html file I kept Next.js to render the maintenance page.

Not the best approach IMO, would love to see Next.js or Vercel offering some solution for this cases.

**Comment by alex-r89** [comment]
Off the back of what @leerob says, I would have a value in your .env file which indicates you are in maintenance mode, ie `MAINTENANCE_MODE=true`

You could then have a `<Maintenance />` component, which you conditionally render in your custom `_app.js` file if that environment variable above is true. If that value is false, just render your usual `<Component {...pageProps} />` or whatever you render. 

Hope that helps, let me know if anything with that solution isn't ideal :)

**Reply by dillionmegida** [reply]
Hi @alex-r89 Thank you so much for this answer. I tried this too, but my site was clustered with the `Maintenance` component not showing.

Here's what I did:

```js
const MyApp: React.FC<AppProps> = ({ Component, pageProps }) => {
  if (process.env.MAINTENANCE) {
    return <Maintenance />;
  }
  return <Component {...pageProps} />;
};

export default MyApp;
```

Please let me know if I'm doing anything wrong here.

**Reply by evgenyfadeev** [reply]
You want a redirect, b/c otherwise the web search may index your pages with the maintenance content.

**Reply by arnemahl** [reply]
To prevent indexing the temporary maintenance page, why not just add `<meta name="robots" content="noindex" />` to the page [`<head>`](https://nextjs.org/docs/api-reference/next/head), @evgenyfadeev?

Can be added in the `Maintenance` component so it won't affect the app under normal circumstances. (And adding the tag will not have permanent effects on indexing, so once the app is rendering normally it can once again be indexed, see https://stackoverflow.com/a/5668421/2054731)

**Comment by mohamedGamalAbuGalala** [comment]
@dillionmegida 
making this environment variable exposed in the client will solve the problem

Just by prefix the env variable by `NEXT_PUBLIC_`

``` typescript
const MyApp: React.FC<AppProps> = ({ Component, pageProps }) => {
  if (process.env.NEXT_PUBLIC_MAINTENANCE === 'true') {
    return <Maintenance />;
  }
  return <Component {...pageProps} />;
};

export default MyApp;
```


Ps. after some testing it needs to be compared with `'true'` to pass the condition.

**Comment by red2678** [comment]
Is there any way to do this in real-time without redeploying the site again?

**Reply by ZRktty** [reply]
Actually redeploying might not be a bad idea!
In this super short tutorial a static site is deployed to replace the whole app.

**Comment by JustFly1984** [comment]
@red2678 Yes, you can use feature toggle/flags, something like LaunchDarkly, or alternatives.

**Comment by ZRktty** [comment]
Found [this ](https://www.jondjones.com/frontend/jamstack/configuring-error-pages-and-maintenance-pages-with-netlify-and-nextjs/)article showing how to deploy a fully static maintenance page replacing the next.js app.
The tutorial shows the process setting up with Netlify but it can work with other hosting solutions as well.
One challenge I had with Netlify, that I had to uninstall the next.js plugins and configure them in toml file to be able to run the deploy job for static files.

**Reply by hoangvu12** [reply]
Can you provide more details. I got the same problem, I uninstalled the plugin and configured it in the `netlify.toml` file. But it failed during the building stage. This is the error it showed.

`Deploy did not succeed: Deploy directory '.next' does not exist`

**Comment by melMass** [comment]
Shouldn't Vercel allow that easily? At an higher level, like it's done on Heroku?

**Reply by 413n** [reply]
Hey @leerob, do you any thought by Vercel on this?
In my use case I think showing a `Maintenance` component is not enough. I would like also to block Next APIs.
I think having the `Maintenance` feature in Vercel that would handle and redirect all requests at a higher level would be great.
I understand that is likely a niche feature, so in case this will not be implemented, do you have any suggestions?
I thought, since I'm on a Turborepo, I could have a `Maintenance app` (basic Next app) and deploy it through the CLI when needed on `Production`. Then with a new deploy on `production` branch a new version will be deployed. Could this work?
I'm not sure about making it work so that the SEO indexes remain as the non-maintenance mode. Thanks!

**Comment by ubinatus** [comment]
Bringing this issue up. Creating a /pages/maintenance.tsx with the `noindex nofollow` meta on the header and then adding the following to the `next.config` file:
```
    redirects() {
      return process.env.NEXT_PUBLIC_SHOW_MAINTENANCE === "1"
        ? [
            {
              source: "/((?!maintenance).*)",
              destination: "/maintenance",
              permanent: false,
            },
          ]
        : [
            {
              source: "/maintenance",
              destination: "/",
              permanent: false,
            },
          ];
    },
```

With this its just a matter of changing your `NEXT_PUBLIC_SHOW_MAINTENANCE` environment variable and rebuilding the project. This will basically redirect al non maintenance pages to the maintenance page. Disabling the maintenance will make the redirection the other way (from maintenance to homepage). 

Pd: make sure you add the `permanent: false` to prevent your browser from caching the redirection ☺️

**Comment by leerob** [comment, marked as answer]
We added a template here for a Next.js application that uses Middleware to toggle maintenance mode: https://vercel.com/templates/next.js/maintenance-page

**Reply by numonedad** [reply]
That's part of the solution. How do we send traffic to it when our backend is down due to ddos?

**Reply by woywro** [reply]
You can always develop a separate branch and promote it to production if necessary. 
Also, take a look at my library https://www.npmjs.com/package/next-maintenance-mode. It's compatible both with edge-config and upstash and includes a caching feature, which is great for saving bandwidth if that's a concern for you.
